File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExpectedExceptionAttributeShouldNotBeUsed.MsTest.vb
Imports System
Imports Microsoft.VisualStudio.TestTools.UnitTesting

Public Class ExceptionTests

    <TestMethod>
    <ExpectedException(GetType(DivideByZeroException))> ' Noncompliant
    Public Sub ExpectedExceptionAttrbutesShouldNotBeUsedOnSub()
        Dim instance As Integer = 42
        instance.ToString()
    End Sub

    <TestMethod>
    <ExpectedException(GetType(DivideByZeroException))> ' Noncompliant
    Public Function ExpectedExceptionAttrbutesShouldNotBeUsedOnFunction() As String
        Dim instance As Integer = 42
        Return instance.ToString()
    End Function

    <TestMethod>
    <ExpectedException(GetType(DivideByZeroException))> ' Compliant - one line
    Public Sub ExpectedExceptionAttrbuteAllowedForSingleLineSub()
        Throw New NotImplementedException()
    End Sub

    <TestMethod>
    <ExpectedException(GetType(DivideByZeroException))> ' Compliant - one line
    Public Function ExpectedExceptionAttrbuteAllowedForSingleLineFunction() As Integer
        Return 42
    End Function

End Class

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExpectedExceptionAttributeShouldNotBeUsed.MsTest.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Tests.Diagnostics
{
    class Program
    {
        [Ignore][TestMethod]
        [ExpectedException(typeof(ArgumentNullException))]  // Noncompliant
        public void TestFoo1()
        {
            var x = true;
            x.ToString();
        }

        [Ignore][TestMethod]
        [ExpectedException(typeof(ArgumentNullException))] // Compliant - one line
        public void TestFoo3()
        {
            new object().ToString();
        }

        [Ignore][TestMethod]
        [ExpectedException(typeof(ArgumentNullException))]  // Compliant - one line
        public string TestFoo5() => new object().ToString();

        [Ignore][TestMethod]
        public void TestFoo7()
        {
            bool callFailed = false;
            try
            {
                //...
            }
            catch (ArgumentNullException)
            {
                callFailed = true;
            }
        }

        [Ignore][TestMethod]
        public void TestWithThrowsAssertation()
        {
            object o = new object();
            Assert.ThrowsException<ArgumentNullException>(() => o.ToString());
        }
    }
}
